src/mlpack/core/cv/simple_cv.hpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * @file simple_cv.hpp
+ * @author Kirill Mishchenko
+ *
+ * A simple cross-validation strategy.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CV_SIMPLE_CV_HPP
+#define MLPACK_CORE_CV_SIMPLE_CV_HPP
+
+#include <mlpack/core/cv/meta_info_extractor.hpp>
+#include <mlpack/core/cv/cv_base.hpp>
+
+namespace mlpack {
+namespace cv {
+
+/**
+ * The class SimpleCV splits data into training and validation sets, runs
+ * training on the training set and evaluates performance on the validation set.
+ *
+ * @tparam MLAlgorithm A machine learning algorithm.
+ * @tparam Metric A metric to assess the quality of a trained model.
+ * @tparam MatType The type of data.
+ * @tparam PredictionsType The type of predictions (should be passed when the
+ *     predictions type is a template parameter in Train methods of
+ *     MLAlgorithm).
+ * @tparam WeightsType The type of weights (should be passed when weighted
+ *     learning is supported, and the weights type is a template parameter in
+ *     Train methods of MLAlgorithm).
+ */
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType = arma::mat,
+         typename PredictionsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType>::PredictionsType,
+         typename WeightsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType,
+                 PredictionsType>::WeightsType>
```

Comment:
1. It is useful to know `MatType`, `PredictionsType` and `WeightsType` before the `Evaluate` call to have an opportunity to prepare datasets for training and validation, so repeated calls (what is the case for hyper-parameter tuning) of Evaluate will have less computational cost. It's not that useful for the simplest case like in SimpleCV, but in general it can make sense (like for k-fold cross-validation). Another reason is to have a similar API to hyper-parameter tuning, where it is more required to pass `MatType`, `PredictionsType` and `WeightsType` into constructor of a tuner in order to have a more nice interface for passing sets of hyper-parameters.
2. Yes, there are situations where `PredictionsType` and `WeightsType` cannot be directly derived from `MLAlgorithm` and `MatType`. A good example is `DecisionTree`, for which `MatType`, `PredictionsType` and `WeightsType` are not specified unless one of Train methods is called. So, if I understand the design of `DecisionTree` in the right way, it is possible to use `arma::Mat<unsigned char>` for `PredictionsType` or `arma::Mat<float>` for `WeightsType`.